### PR TITLE
[runtime env] Fix bug where options (e.g. `--extra-index-url`) could not be specified in `requirements.txt`

### LINKED ISSUE #22056: [Bug] Python: --extra-index-url does not work in requirements file anymore
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

Up until Ray 1.9.0, I could provide a call to `ray.init(address=head_address, runtime_env={'pip': requirements_file})` with a `requirements_file` that preprended the line `--extra-index-url INDEX` within the file to provide an alternative index (see https://stackoverflow.com/a/2477610)

Starting with Ray 1.10.0, I am now getting an error:
```
Traceback (most recent call last):
  File "/opt/launch_summit", line 56, in <module>
    ray.init(address=head_address, runtime_env={'pip': req_path})
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/worker.py", line 977, in init
    connect(
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/worker.py", line 1523, in connect
    serialized_job_config = job_config.serialize()
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/job_config.py", line 51, in serialize
    return self.get_proto_job_config().SerializeToString()
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/job_config.py", line 108, in get_proto_job_config
    parsed_env, eager_install = self._validate_runtime_env()
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/job_config.py", line 92, in _validate_runtime_env
    return ParsedRuntimeEnv(self.runtime_env), eager_install
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/runtime_env/validation.py", line 345, in __init__
    validated_option_val = validate_fn(option_val)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/runtime_env/validation.py", line 179, in parse_and_validate_pip
    result = _rewrite_pip_list_ray_libraries(pip_list)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/ray/_private/runtime_env/validation.py", line 115, in _rewrite_pip_list_ray_libraries
    requirement = Requirement.parse(specifier)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/pkg_resources/__init__.py", line 3139, in parse
    req, = parse_requirements(s)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/pkg_resources/__init__.py", line 3084, in parse_requirements
    yield Requirement(line)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/pkg_resources/__init__.py", line 3094, in __init__
    super(Requirement, self).__init__(requirement_string)
  File "/home/ray/anaconda3/lib/python3.8/site-packages/pkg_resources/_vendor/packaging/requirements.py", line 100, in __init__
    raise InvalidRequirement(
pkg_resources.extern.packaging.requirements.InvalidRequirement: Parse error at "'--extra-'": Expected W:(abcd...) 
```

Any ideas on what has changed and how I can get this behavior back on the latest Ray?

### Versions / Dependencies

Python 3.8.6
Ray 1.10.0 or higher
Ubuntu 18.04.5 LTS

### Reproduction script

requirements.txt:
```
--extra-index-url PRIVATE_PACKAGE_INDEX
private-package==1.0.0
```

script.py
```
import ray
ray.init(runtime_env={'pip': 'requirements.txt'})
```

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
In https://github.com/ray-project/ray/pull/20341 the behavior of `pip` was changed to install the specified packages in the existing environment rather than in a new environment.  This posed a problem when specifying Ray libraries like "ray[serve]" in the `pip` field, because the installer would install Ray at runtime and this new Ray would take precedence over the Ray existing on the cluster.  This could cause version mismatch issues.  Skipping some details, the approach taken in the that PR was essentially to parse the `pip` list and remove Ray. 

However not every line in a `pip` `requirements.txt` file is a requirements specifier; a line can also just specify options, like `--extra-index-url my-index-url.com`. 
 This caused the parsing library to raise an exception when trying to parse the line.  This PR fixes this by catching the exception and skipping the line in this case, since it's not a line that specifies `ray` and that's all we're looking for when parsing.

## Related issue number

Closes #22056 
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
@edoakes Thoughts on whether or not we should tag this with `v1.11-pick`?  You would be the approver.

In favor of picking
- It's a regression
- This feature is necessary to use private package repositories, which are a common requirement at companies.
- Low risk of affecting release tests

Against picking:
- I'm not sure if the issue meets the bar for P0, and we should only cherry pick fixes to P0 issues.
- You can work around this by using the `conda` field of runtime_env instead (using the `pip` subfield), though then you don't get the benefit of the `pip` incremental install.

**Comment by edoakes** [comment]
@architkulkarni this looks like a P0 issue to me (there is a known user workflow that is completely broken by this bug), so I'd say we should pick. This change also looks low-risk to me.

File: python/ray/_private/runtime_env/validation.py
Comment (edoakes) [inline python/ray/_private/runtime_env/validation.py:145]: just to be clear, these options will still be passed through to the final call to create the env, right?

**Review by edoakes — APPROVED** [review]
LGTM assuming my assumption/clarification is correct

File: python/ray/_private/runtime_env/validation.py
Comment (architkulkarni) [reply python/ray/_private/runtime_env/validation.py:145]: Oh wow, good catch.  Let me fix this and fix the test so that it catches this.

Comment (architkulkarni) [reply python/ray/_private/runtime_env/validation.py]: Fixed

**Comment by architkulkarni** [comment]
@simon-mo could you please approve this PR?  We need your approval, per the following:
> Merging a cherry-pick PR would require 2 approvals -- one committer and the TL owning the component

**Review by simon-mo — APPROVED** [review]
stamp

**Comment by architkulkarni** [comment]
@edoakes I had to make another change to get it to work, could you please review it?

Basically, it's not the case that `pip install -r requirements.txt` is equivalent to calling `pip install` on each line (which is what we currently do), so we need to dump the lines into a `requirements.txt` file.  This is what we currently do with the conda dict, for example.

I think for this reason, it makes more sense to use a single `str` in the runtime env protobuf for the pip field, rather than a list of `str` called "packages" as we do currently.  I can make a followup PR for this.

File: python/ray/_private/runtime_env/pip.py
Comment (edoakes) [inline python/ray/_private/runtime_env/pip.py:42]:
```suggestion
        with open(pip_requirements_file, "w") as file:
            print("\n".join(pip_list), file=file)
```
nit: more concise

File: python/ray/_private/runtime_env/validation.py
Comment (edoakes) [inline python/ray/_private/runtime_env/validation.py:119]: thanks for adding

**Review by edoakes — APPROVED** [review]

**Comment by edoakes** [comment]
no need to address the nit if it slows down merging

**Comment by edoakes** [comment]
Some suspicious-looking test failures (`test_client_proxy`)

**Pull request merged**
